medallion/test/config/test_config_loading.py

```diff
@@ -0,0 +1,253 @@
+import re
+from unittest import mock
+
+import pytest
+import pytest_subtests  # noqa: F401
+
+import medallion.config as m_cfg
+
+
+def test_load_config_defaults():
+    """
+    Ensure that `conf_file` and `conf_dir` arguments have expected defaults.
+
+    We do this since we don't want to attempt to create those files for parsing
+    in any of our actual functional tests below.
+    """
+    assert m_cfg.load_config.__defaults__ == (
+        m_cfg.DEFAULT_CONFFILE, m_cfg.DEFAULT_CONFDIR,
+    )
+
+
+def test_load_config_default_file_missing(subtests, tmp_path):
+    """
+    Confirm that nothing bad happens if the default config file is missing.
+    """
+    with mock.patch(
+        "medallion.config.DEFAULT_CONFFILE", str(tmp_path / "medallion.conf"),
+    ) as mock_conf_file_default:
+        with subtests.test(msg="conf_dir omitted"):
+            config = m_cfg.load_config(conf_file=mock_conf_file_default)
+            assert config == {}
+        with subtests.test(msg="conf_dir is None"):
+            config = m_cfg.load_config(
+                conf_file=mock_conf_file_default, conf_dir=None,
+            )
+            assert config == {}
+
+
+def test_load_config_non_default_file_missing(subtests, tmp_path):
+    """
+    Confirm an exception is raised if a non-default config file is missing.
+    """
+    conf_file = tmp_path / "missing.conf"
+    exc_patt = re.escape(repr(str(conf_file)))
+    with pytest.raises(FileNotFoundError, match=exc_patt):
+        m_cfg.load_config(conf_file=conf_file)
+
+
+def test_load_config_default_dir_missing(subtests, tmp_path):
+    """
+    Confirm that nothing bad happens if the default config dir is missing.
+    """
+    with mock.patch(
+        "medallion.config.DEFAULT_CONFDIR", str(tmp_path / "medallion.d"),
+    ) as mock_conf_dir_default:
+        with subtests.test(msg="conf_file omitted"):
+            config = m_cfg.load_config(conf_dir=mock_conf_dir_default,)
+            assert config == {}
+        with subtests.test(msg="conf_file is None"):
+            config = m_cfg.load_config(
+                conf_dir=mock_conf_dir_default, conf_file=None,
+            )
+            assert config == {}
+
+
+def test_load_config_non_default_dir_missing(tmp_path):
+    """
+    Confirm an exception is raised if a non-default config dir is missing.
+    """
+    conf_dir = tmp_path / "missing.d"
+    exc_patt = re.escape(repr(str(conf_dir)))
+    with pytest.raises(FileNotFoundError, match=exc_patt):
+        m_cfg.load_config(conf_dir=conf_dir)
+
+
+@pytest.mark.parametrize(
+    "confdata", ({}, {"foo": "bar"}, {"foo": {"bar": ["baz", ]}}, )
+)
+def test_load_config_json_objects(confdata, subtests, tmp_path):
+    """
+    Confirm that config files get loaded correctly.
+    """
+    conf_file = tmp_path / "medallion.conf"
+    json.dump(confdata, conf_file.open("w"))
+    with subtests.test(msg="Used as config file"):
+        assert m_cfg.load_config(conf_file=conf_file) == confdata
```

Comment:
Because this function checks environment variables, if variables are set when running the tests, it will cause a lot of them to fail. Is there a way to unset environment variables at test execution time, then reset them afterwards?

Reply:
I believe there is a pytest monkey_patch fixture which might do the job: https://docs.pytest.org/en/6.2.x/monkeypatch.html#monkeypatching-environment-variables

I'll look into some best practice for this - I think patching the `os.environ` dict might be reasonable but if that's not easy/recommended, we could potentially mock `environ_config` to pull from an alternative default source of variables.

Reply:
You could also get, delete, then reset them with a pytest fixture, though that's definitely more rough than using something like monkeypatch, if it works.

Reply:
Done, I sanity checked that the amended diff works with a poisonous envvar by applying the following patch. The previous diff would have all sorts of failures with the same envvar.

```diff
diff --git a/tox.ini b/tox.ini
index fb9be89..2d714ad 100644
--- a/tox.ini
+++ b/tox.ini
@@ -12,6 +12,8 @@ deps =
     responses
     pymongo
     pyjwt
+setenv =
+    MEDALLION_TAXII_MAX_PAGE_SIZE = 1337
 commands =
     pytest --cov=medallion medallion/test/ --cov-report term-missing
 
```